File: Collections/CreateInstanceDictionary.vb
﻿Namespace Common

    Public NotInheritable Class CreateInstanceDictionary

        Private Sub New()
            Throw New NotSupportedException()
        End Sub

        Public Shared Function Create(Of TKey, TValue)(ByVal Dic As IDictionary(Of TKey, TValue), ByVal Creator As Func(Of TKey, TValue)) As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Dic, Creator)
        End Function

        Public Shared Function Create(Of TKey, TValue)(ByVal Creator As Func(Of TKey, TValue)) As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Creator)
        End Function

        <Obsolete("Use other overloads.", True)>
        Public Shared Function Create(Of TKey, TValue)(ByVal Dic As IDictionary(Of TKey, TValue), ByVal Creator As Func(Of TValue)) As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Dic, Function(K) Creator.Invoke())
        End Function

        <Obsolete("Use other overloads.", True)>
        Public Shared Function Create(Of TKey, TValue)(ByVal Creator As Func(Of TValue)) As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Function(K) Creator.Invoke())
        End Function

        Public Shared Function Create(Of TKey, TValue As New)(ByVal Dic As IDictionary(Of TKey, TValue)) As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Dic, Function() New TValue())
        End Function

        Public Shared Function Create(Of TKey, TValue As New)() As CreateInstanceDictionary(Of TKey, TValue)
            Return New CreateInstanceDictionary(Of TKey, TValue)(Function() New TValue())
        End Function

    End Class

    Public Class CreateInstanceDictionary(Of TKey, TValue)
        Inherits BaseDictionary(Of TKey, TValue)

        Public Sub New(ByVal Dic As IDictionary(Of TKey, TValue), ByVal Creator As Func(Of TKey, TValue))
            Me.Dic = Dic
            Me.Creator = Creator
        End Sub

        Public Sub New(ByVal Creator As Func(Of TKey, TValue))
            Me.New(New Dictionary(Of TKey, TValue)(), Creator)
        End Sub

        Public Overrides Sub Clear()
            Me.Dic.Clear()
        End Sub

        Public Overrides ReadOnly Property Count As Integer
            Get
                Return Me.Dic.Count
            End Get
        End Property

        Public Overrides Sub Add(key As TKey, value As TValue)
            Me.Dic.Add(key, value)
        End Sub

        Public Overrides Function ContainsKey(key As TKey) As Boolean
            Return Me.Dic.ContainsKey(key)
        End Function

        Default Public Overrides Property Item(key As TKey) As TValue
            Get
                Dim V As TValue
                If Not Me.Dic.TryGetValue(key, V) Then
                    V = Me.Creator.Invoke(key)
                    Me.Dic.Add(key, V)
                End If
                Return V
            End Get
            Set(ByVal value As TValue)
                Me.Dic.Item(key) = value
            End Set
        End Property

        Public Overrides ReadOnly Property Keys As ICollection(Of TKey)
            Get
                Return Me.Dic.Keys
            End Get
        End Property

        Public Overrides Function Remove(key As TKey) As Boolean
            Return Me.Dic.Remove(key)
        End Function

        Public Overrides Function TryGetValue(key As TKey, ByRef value As TValue) As Boolean
            Return Me.Dic.TryGetValue(key, value)
        End Function

        Public Overrides ReadOnly Property Values As ICollection(Of TValue)
            Get
                Return Me.Dic.Values
            End Get
        End Property

        Public Function GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Return Me.Dic.GetEnumerator()
        End Function

        Protected Overrides Function IEnumerator_1_GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Return Me.GetEnumerator()
        End Function

        Private ReadOnly Creator As Func(Of TKey, TValue)
        Private ReadOnly Dic As IDictionary(Of TKey, TValue)

    End Class

End Namespace

File: Collections/CreateInstanceDictionary.cs
using System;
using System.Collections.Generic;

namespace Ks.Common
{
    public static class CreateInstanceDictionary
    {

        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>(IDictionary<TKey, TValue> Dic, Func<TKey, TValue> Creator)
        {
            return new CreateInstanceDictionary<TKey, TValue>(Dic, Creator);
        }

        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>(Func<TKey, TValue> Creator)
        {
            return new CreateInstanceDictionary<TKey, TValue>(Creator);
        }

        [Obsolete("Use other overloads.", true)]
        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>(IDictionary<TKey, TValue> Dic, Func<TValue> Creator)
        {
            return new CreateInstanceDictionary<TKey, TValue>(Dic, K => Creator.Invoke());
        }

        [Obsolete("Use other overloads.", true)]
        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>(Func<TValue> Creator)
        {
            return new CreateInstanceDictionary<TKey, TValue>(K => Creator.Invoke());
        }

        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>(IDictionary<TKey, TValue> Dic) where TValue : new()
        {
            return new CreateInstanceDictionary<TKey, TValue>(Dic, K => new TValue());
        }

        public static CreateInstanceDictionary<TKey, TValue> Create<TKey, TValue>() where TValue : new()
        {
            return new CreateInstanceDictionary<TKey, TValue>(K => new TValue());
        }
    }

    public class CreateInstanceDictionary<TKey, TValue> : BaseDictionary<TKey, TValue>
    {
        public CreateInstanceDictionary(IDictionary<TKey, TValue> Dic, Func<TKey, TValue> Creator)
        {
            this.Dic = Dic;
            this.Creator = Creator;
        }

        public CreateInstanceDictionary(Func<TKey, TValue> Creator) : this(new Dictionary<TKey, TValue>(), Creator)
        {
        }

        public override void Clear()
        {
            this.Dic.Clear();
        }

        public override int Count
        {
            get
            {
                return this.Dic.Count;
            }
        }

        public override void Add(TKey key, TValue value)
        {
            this.Dic.Add(key, value);
        }

        public override bool ContainsKey(TKey key)
        {
            return this.Dic.ContainsKey(key);
        }

        public override TValue this[TKey key]
        {
            get
            {
                if (!this.Dic.TryGetValue(key, out var V))
                {
                    V = this.Creator.Invoke(key);
                    this.Dic.Add(key, V);
                }
                return V;
            }
            set
            {
                this.Dic[key] = value;
            }
        }

        public override ICollection<TKey> Keys
        {
            get
            {
                return this.Dic.Keys;
            }
        }

        public override bool Remove(TKey key)
        {
            return this.Dic.Remove(key);
        }

        public override bool TryGetValue(TKey key, out TValue value)
        {
            return this.Dic.TryGetValue(key, out value);
        }

        public override ICollection<TValue> Values
        {
            get
            {
                return this.Dic.Values;
            }
        }

        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
        {
            return this.Dic.GetEnumerator();
        }

        protected override IEnumerator<KeyValuePair<TKey, TValue>> _GetEnumerator()
        {
            return this.GetEnumerator();
        }

        private readonly Func<TKey, TValue> Creator;
        private readonly IDictionary<TKey, TValue> Dic;
    }
}
